File: advent-of-code-2024/day4.py
with open('input') as f:
    grid = [[ch for ch in line.strip()] for line in f.readlines()]

part1 = 0
res = ['XMAS', 'SAMX']
for i in range(0, len(grid)):
    for j in range(0, len(grid[0])):
        if grid[i][j] in ['X', 'S']:
            if j + 3 < len(grid[0]):
                h = grid[i][j] + grid[i][j+1] + grid[i][j+2] + grid[i][j+3] 
                if h in res: part1 += 1
            if i + 3 < len(grid):
                v = grid[i][j] + grid[i+1][j] + grid[i+2][j] + grid[i+3][j]
                if v in res: part1 += 1

            if i + 3 < len(grid) and j + 3 < len(grid[0]):
                d1 = grid[i][j] + grid[i+1][j+1] + grid[i+2][j+2] + grid[i+3][j+3] 
                if d1 in res: part1 += 1
            if i + 3 < len(grid) and j - 3 >= 0:
                d2 = grid[i][j] + grid[i+1][j-1] + grid[i+2][j-2] + grid[i+3][j-3] 
                if d2 in res: part1 += 1

print(part1)

part2 = 0
for i in range(1, len(grid)-1):
    for j in range(1, len(grid[0])-1):
        if grid[i][j] == 'A':
            if grid[i-1][j-1] == 'M' and grid[i-1][j+1] == 'M' and grid[i+1][j-1] == 'S' and grid[i+1][j+1] == 'S': part2 += 1
            elif grid[i-1][j-1] == 'S' and grid[i-1][j+1] == 'M' and grid[i+1][j-1] == 'S' and grid[i+1][j+1] == 'M': part2 += 1
            elif grid[i-1][j-1] == 'S' and grid[i-1][j+1] == 'S' and grid[i+1][j-1] == 'M' and grid[i+1][j+1] == 'M': part2 += 1
            elif grid[i-1][j-1] == 'M' and grid[i-1][j+1] == 'S' and grid[i+1][j-1] == 'M' and grid[i+1][j+1] == 'S': part2 += 1

print(part2)






        

File: advent-of-code-2024/cpp/day4.cpp
# include <iostream>
# include <fstream>
# include <vector>

using namespace std;

int main(){
    ifstream file("../input");

    vector<string> grid;
    string str;

    while(file >> str){
        grid.push_back(str);
    }

    int part1 = 0;
    int dir[8][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}, {-1, -1}, {-1, 1}, {1, -1}, {1, 1}};

    for(int i=0; i<grid.size(); i++){
        for(int j=0; j<grid[0].length(); j++){
            if(grid[i][j] == 'X'){
                if(j + 3 < grid[0].length() && grid[i].substr(j, 4) == "XMAS") part1++;
                if(j - 3 >= 0 && grid[i].substr(j-3, 4) == "SAMX") part1++;
                if(i + 3 < grid.size() && grid[i+1][j] == 'M' && grid[i+2][j] == 'A' && grid[i+3][j] == 'S') part1++;
                if(i - 3 >= 0 && grid[i-1][j] == 'M' && grid[i-2][j] == 'A' && grid[i-3][j] == 'S') part1++;

                if(i - 3 >= 0 && j - 3 >= 0 && grid[i-1][j-1] == 'M' && grid[i-2][j-2] == 'A' && grid[i-3][j-3] == 'S') part1++;
                if(i - 3 >= 0 && j + 3 < grid[0].length() && grid[i-1][j+1] == 'M' && grid[i-2][j+2] == 'A' && grid[i-3][j+3] == 'S') part1++;
                if(i + 3 < grid.size() && j - 3 >= 0 && grid[i+1][j-1] == 'M' && grid[i+2][j-2] == 'A' && grid[i+3][j-3] == 'S') part1++;
                if(i + 3 < grid.size() && j + 3 < grid[0].length() && grid[i+1][j+1] == 'M' && grid[i+2][j+2] == 'A' && grid[i+3][j+3] == 'S') part1++;

            }
        }
    }

    int part2 = 0;
    for(int i=1; i<grid.size()-1; i++){
        for(int j=1; j<grid[0].length()-1; j++){
            if(grid[i][j] == 'A'){
                if(grid[i-1][j-1] == 'M' && grid[i-1][j+1] == 'M' && grid[i+1][j-1] == 'S' && grid[i+1][j+1] == 'S') part2++;
                if(grid[i-1][j-1] == 'S' && grid[i-1][j+1] == 'S' && grid[i+1][j-1] == 'M' && grid[i+1][j+1] == 'M') part2++;
                if(grid[i-1][j-1] == 'M' && grid[i-1][j+1] == 'S' && grid[i+1][j-1] == 'M' && grid[i+1][j+1] == 'S') part2++;
                if(grid[i-1][j-1] == 'S' && grid[i-1][j+1] == 'M' && grid[i+1][j-1] == 'S' && grid[i+1][j+1] == 'M') part2++;
            }
        }
    }

    cout << "part 1: " << part1 << endl;
    cout << "part 2: " << part2 << endl;
}
